File: strings/reverse_substr_between_each_pair_of_parantheses.go
/*
  You are given a string s that consists of lower case English letters and brackets.
  Reverse the strings in each pair of matching parentheses, starting from the innermost one.
  Your result should not contain any brackets.

  Example:
    Input: s = "(abcd)"
    Output: "dcba"
*/
func reverseParentheses(s string) string {
    st := []rune{};

    for _, c := range s {
        if c != ')' {
            st = append(st, c);
        } else {
            q := []rune{};
            for len(st) > 0 && st[len(st) - 1] != '(' {
                q = append(q, st[len(st) - 1]);
                st = st[:len(st) - 1];
            }
            if len(st) > 0 {
                st = st[:len(st) - 1];
            }
            for _, c1 := range q {
                st = append(st, c1);
            }
        }
    }

    return string(st);
}

File: strings/reverse_substr_between_each_pair_of_parantheses.cs
/*
  You are given a string s that consists of lower case English letters and brackets.
  Reverse the strings in each pair of matching parentheses, starting from the innermost one.
  Your result should not contain any brackets.

  Example:
    Input: s = "(abcd)"
    Output: "dcba"
*/
public class Solution {
    public string ReverseParentheses(string s) {
        Stack<char> st = [];
        Queue<char> q = [];

        foreach (char c in s) {
            if (c != ')') st.Push(c);
            else {
                while (st.TryPop(out char c1) && c1 != '(') q.Enqueue(c1);
                while (q.TryDequeue(out char c2)) st.Push(c2);
            }
        }

        Span<char> result = stackalloc char[st.Count];
        for (int i = st.Count - 1; i >= 0; --i) {
            result[i] = st.Pop();
        }
        return new string(result);
    }
}
